### PR TITLE
Add bazel, update supporting documentation

### PR DESCRIPTION
**Pull request by justaugustus**
**What this PR does / why we need it**:
- Update .gitignore, Makefile
- Update documentation
- Add .bazelrc, BUILD.bazel, scripts/hack dir
- Generate bazel files
- Update vendor

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
Add bazel for building capz
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,123 +1,99 @@
-# Kubernetes Cluster API Azure Provider  [![Go Report Card](https://goreportcard.com/badge/sigs.k8s.io/cluster-api-provider-azure)](https://goreportcard.com/report/kubernetes-sigs/cluster-api-provider-azure)
+# Kubernetes Cluster API Provider Azure
 
-## Getting Started
+[![Go Report Card](https://goreportcard.com/badge/sigs.k8s.io/cluster-api-provider-azure)](https://goreportcard.com/report/sigs.k8s.io/cluster-api-provider-azure)
```
Comment (tariq1890) [inline README.md]:
```suggestion
[![Go Report Card](https://goreportcard.com/badge/kubernetes-sigs/cluster-api-provider-azure)](https://goreportcard.com/report/kubernetes-sigs/cluster-api-provider-azure)
```

File: Makefile [hunk 2]
```diff
@@ -12,40 +12,221 @@
...
+
+.PHONY: reset-bazel
+reset-bazel: ## Deep cleaning for bazel
+	bazel clean --expunge
+
+cmd/clusterctl/examples/azure/out:
+	./cmd/clusterctl/examples/azure/generate-yaml.sh
+
+# TODO: Uncomment once clusterazureadm exists
+#cmd/clusterctl/examples/azure/out/credentials: cmd/clusterctl/examples/azure/out ## Generate k8s secret for AWS credentials
```
Comment (tariq1890) [inline Makefile]:
```suggestion
#cmd/clusterctl/examples/azure/out/credentials: cmd/clusterctl/examples/azure/out ## Generate k8s secret for Azure credentials
```

File: Makefile [hunk 3]
```diff
@@ -12,40 +12,221 @@
...
+
+# TODO(vincepri): This should move to rebuild Bazel binaries once every
+# make target uses Bazel bins to run operations.
+.PHONY: binaries-dev
+binaries-dev: ## Builds and installs the binaries on the local GOPATH
+	go get -v ./...
+	go install -v ./...
+
+.PHONY: create-cluster
+create-cluster: ## Create a Kubernetes cluster on AWS using examples
```
Comment (tariq1890) [inline Makefile]:
```suggestion
create-cluster: ## Create a Kubernetes cluster on Azure using examples
```

File: README.md (hunk 1 above)
Comment (justaugustus) [reply README.md]: This doesn't need to be fixed. That's the correct import path.

Comment (tariq1890) [reply README.md]: Unfortunately goreportcard requires the absolute links for the github permalinks inside of the lint reports to work.

**Comment by justaugustus** [comment]
(Closing to hopefully clear the failing tests)
/close

**Comment by justaugustus** [comment]
New pre-submit tests are merged: kubernetes/test-infra#11024
They're passing on this PR, but of course the old ones are going to continue to fail, so I'll need to merge this manually.

There are continued improvements that can be made to the documentation, but this needs to merge ASAP to ensure we're getting good signal on new PRs coming in.

@tariq1890 / @soggiest -- please review, `/lgtm`, and `/approve` when you have a chance, then I'll manually merge this.

File: docs/development.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+`make copy-genmocks`
+
+<!-- References -->
+
+[jq]: https://stedolan.github.io/jq/download/
+[image_pull_secrets]: https://kubernetes.io/docs/concepts/containers/images/#specifying-imagepullsecrets-on-a-pod
+[gettext]: https://www.gnu.org/software/gettext/
+[kind]: https://sigs.k8s.io/kind
+[azure_cli]: https://docs.azure.amazon.com/cli/latest/userguide/installing.html
```
Comment (tariq1890) [inline docs/development.md]:
```suggestion
[azure_cli]: https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest
```

File: docs/accessing-instances.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+```bash
+export CLUSTER_SSH_KEY=$HOME/.ssh/cluster-api-provider-azure
+```
+
+### Obtain public IP address of the bastion node
+
+> Your credentials must let you query EC2 API
+
+```bash
+export BASTION_HOST=$(azure ec2 describe-instances --filter='Name=tag:Name,Values=<CLUSTER_NAME>-bastion' \
```
Comment (tariq1890) [inline docs/accessing-instances.md]: I see that this has a lot of amazon specific stuff. I understand that the intent is to fix these in a follow-up PR. However, from past experiences, it's much harder to track all of the doc debt after merge. Is it possible to scope this in another PR?

Comment (tariq1890) [reply docs/accessing-instances.md]: We can exclude these docs which have a lot of aws content basically.

Comment (justaugustus) [reply docs/accessing-instances.md]: Meant to remove that file. Thanks for catching. Just fixed!

File: docs/clusterctl.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
+# clusterctl
```
Comment (tariq1890) [inline docs/clusterctl.md]: This is some really helpful documentation :)

**Comment by tariq1890** [comment]
Thanks very much @justaugustus :)

**Pull request merged**